encoding/ccf/traverse_value.go

```diff
@@ -0,0 +1,232 @@
+/*
+ * Cadence - The resource-oriented smart contract programming language
+ *
+ * Copyright Dapper Labs, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package ccf
+
+import (
+	"sort"
+
+	"github.com/onflow/cadence"
+)
+
+type compositeTypes struct {
+	ids           ccfTypeIDByCadenceType
+	abstractTypes map[string]bool
+	types         []cadence.Type
+}
+
+// compositeTypesFromValue returns all composite/interface types for value v.
+// Returned types are sorted unique list of static and runtime composite/interface types.
+// NOTE: nested composite/interface types are included in the returned types.
+func compositeTypesFromValue(v cadence.Value) ([]cadence.Type, ccfTypeIDByCadenceType) {
+	ct := &compositeTypes{
+		ids:           make(ccfTypeIDByCadenceType),
+		abstractTypes: make(map[string]bool),
+		types:         make([]cadence.Type, 0, 1),
+	}
+
+	// Traverse v to get all unique:
+	// - static composite types
+	// - static interface types
+	// - runtime composite types
+	// - runtime interface types
+	ct.traverseValue(v)
+
+	if len(ct.ids) < 2 {
+		// No need to reassign ccf id, nor sort types.
+		return ct.types, ct.ids
+	}
+
+	// Sort Cadence types by Cadence type ID.
+	sort.Sort(bytewiseCadenceTypeInPlaceSorter(ct.types))
+
+	// Assign sorted array index as local ccf ID.
+	for i, typ := range ct.types {
+		ct.ids[typ.ID()] = ccfTypeID(i)
+	}
+
+	return ct.types, ct.ids
+}
+
+func (ct *compositeTypes) traverseValue(v cadence.Value) {
+
+	// Traverse type for composite/interface types.
+	checkRuntimeType := ct.traverseType(v.Type())
+
+	if !checkRuntimeType {
+		// Return without traversing value for runtime types.
+		return
+	}
+
+	// Traverse v's elements for runtime types.
+	// Note: don't need to traverse fields of cadence.Enum
+	// because enum's field is an integer subtype.
+	switch x := v.(type) {
```

Comment:
Can shadow the variable name here:

```suggestion
	switch v := v.(type) {
```

Reply:
Done in commit 926661cbc672ec8711015c5a097586a81e2f3d2b.

While at it, also applied this suggestion to other places.